apps/cow-tools/src/lib/transactionFactory.ts

```diff
@@ -0,0 +1,121 @@
+import { Address } from "@bleu-fi/utils";
+import { BaseTransaction } from "@gnosis.pm/safe-apps-sdk";
+import { erc20ABI } from "@wagmi/core";
+import { encodeFunctionData } from "viem";
+import { goerli } from "viem/chains";
+
+import { milkmanAbi } from "#/lib/abis/milkman";
+
+import {
+  encodePriceCheckerData,
+  PRICE_CHECKERS,
+  priceCheckerInfoMapping,
+} from "./priceCheckers";
+
+// Milkman's address is the same for all chains supported chains (Gnosis, etc...)
+export const MILKMAN_ADDRESS = "0x11C76AD590ABDFFCD980afEC9ad951B160F02797";
+
+export enum TRANSACTION_TYPES {
+  ERC20_APPROVE = "ERC20_APPROVE",
+  MILKMAN_ORDER = "MILKMAN_ORDER",
+}
+export interface BaseArgs {
+  type: TRANSACTION_TYPES;
+}
+
+export interface ERC20ApproveArgs extends BaseArgs {
+  type: TRANSACTION_TYPES.ERC20_APPROVE;
+  tokenAddress: Address;
+  spender: Address;
+  amount: bigint;
+}
+
+export interface MilkmanOrderArgs extends BaseArgs {
+  type: TRANSACTION_TYPES.MILKMAN_ORDER;
+  tokenAddressToSell: Address;
+  tokenAddressToBuy: Address;
+  toAddress: Address;
+  amount: bigint;
+  priceChecker: PRICE_CHECKERS;
+  args: bigint[];
+}
+
+interface ITransaction<T> {
+  createRawTx(args: T): BaseTransaction;
+}
+
+class ERC20ApproveRawTx implements ITransaction<ERC20ApproveArgs> {
+  createRawTx({
+    tokenAddress,
+    spender,
+    amount,
+  }: ERC20ApproveArgs): BaseTransaction {
+    return {
+      to: tokenAddress,
+      value: "0",
+      data: encodeFunctionData({
+        abi: erc20ABI,
+        functionName: "approve",
+        args: [spender, amount],
+      }),
+    };
+  }
+}
+
+class MilkmanOrderRawTx implements ITransaction<MilkmanOrderArgs> {
+  createRawTx({
+    tokenAddressToSell,
+    tokenAddressToBuy,
+    toAddress,
+    amount,
+    priceChecker,
+    args,
+  }: MilkmanOrderArgs): BaseTransaction {
+    const priceCheckerInfo = priceCheckerInfoMapping[priceChecker];
+    const priceCheckerAddress = priceCheckerInfo.addresses[goerli.id];
+    const priceCheckerData = encodePriceCheckerData(priceChecker, args);
```

Comment:
I wasn't sure this should be `[args]` as before because this was already being passed as an array to this function. I couldn't run a transaction locally to test =/